flocker/node/agents/ebs.py

```diff
@@ -243,74 +383,125 @@ def _blockdevicevolume_from_ebs_volume(ebs_volume):
-                     start_status,
-                     transient_status,
-                     end_status):
+def _get_ebs_volume_state(volume):
+    """
+    Fetch input EBS volume's latest state from backend.
+
+    :param boto.ec2.volume volume: Volume that needs state update.
+
+    :returns: EBS volume with latest state known to backend.
+    :rtype: boto.ec2.volume
+
+    """
+    return volume.update()
+
+
+def _should_finish(operation, volume, update, start_time,
+                   timeout=VOLUME_STATE_CHANGE_TIMEOUT):
+    """
+    Helper function to determine if wait for volume's state transition
+    resulting from given operation is over.
+    The method completes if volume reached expected end state, or, failed
+    to reach expected end state, or we timed out waiting for the volume to
+    reach expected end state.
+
+    :param NamedConstant operation: Operation performed on given volume.
+    :param boto.ec2.volume volume: Target volume of given operation.
+    :param method update: Method to use to check volume state.
+    :param float start_time: Time when operation was executed on volume.
+    :param int timeout: Time, in seconds, to wait for volume to reach expected
+        destination state.
+
+    :returns: True or False indicating end of wait for volume state transition.
+    :rtype: bool
+    """
+    state_flow = VOLUME_STATE_TABLE.table[operation]
+    start_state = state_flow.start_state.value
+    transient_state = state_flow.transient_state.value
+    end_state = state_flow.end_state.value
+    sets_attach = state_flow.sets_attach
+    unsets_attach = state_flow.unsets_attach
+
+    if time.time() - start_time > timeout:
+        # We either:
+        # 1) Timed out waiting to reach ``end_status``, or,
+        # 2) Reached an unexpected status (state change resulted in error), or,
+        # 3) Reached ``end_status``, but ``end_status`` comes with
+        #    attach data, and we timed out waiting for attach data.
+        # Raise a ``TimeoutException`` in all cases.
+        raise TimeoutException(unicode(volume.id), operation, start_state,
+                               transient_state, end_state, volume.status)
+
+    try:
+        update(volume)
+    except EC2ResponseError as e:
+        # If AWS cannot find the volume, raise ``UnknownVolume``.
+        # (http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html
+        # for error details).
+        if e.code == u'InvalidVolume.NotFound':
+            raise UnknownVolume(volume.id)
+
+    if volume.status == end_state:
+        # If end state for the volume comes with attach data,
+        # declare success only upon discovering attach data.
+        if sets_attach:
+            if (volume.attach_data is not None and
+                    (volume.attach_data.device != '' and
+                     volume.attach_data.instance_id != '')):
+                return True
+            else:
+                return False
```

Comment:
I don't think the `if`/`else` here adds to the readability. Please change to just return the condition, e.g.

```
        return (volume.attach_data is not None and
                (volume.attach_data.device != '' and
                 volume.attach_data.instance_id != ''))
```